### ISSUE TITLE
[BUG] Gradient tape *inside* tf.function broken for tf.Variable argument.

### ISSUE THREAD

**Issue by jonas-eschle** [tensorflow/tensorflow#57365]
### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.7, 2.9, nightly

### Custom Code

Yes

### OS Platform and Distribution

Ubuntu 22.04

### Python version

3.1)

### Current Behaviour?

```shell
tf.function (since TF 2.7 I believe) now remembers the storage of the variable and treats them as tensors (using the `__tf_retrace__` mechanism). This works fine whenever the value of the variable is needed and not it's identity. However, when taking the gradient, the identity matters, not the value.

This means when decorating a function that takes the gradient of y with respect to x (x being a parameter here), it traces the function, uses the value of the TensorLike object and **creates the gradient with respect to the parameter's identity (watching the parameter)**.
Calling the same function with a different parameter won't retrace (!) and just use the value, however since it doesn't retrace, it won't re-calculate the gradient with respect to the *identity* of the new variable given as argument but instead use the remembered gradient.

Basically, the problem arises since tf.function and tensor-like objects function functionally (i.e. gradient with respect to argument one, two etc) and tf.Variable, that are global statebased objects. Using a statebased object as a "functional" object breaks the correct behavior of GradientTape.

## Bug or feature?

I do understand the merits of this in a function that doesn't take gradients. However, it breaks for the logic of the gradient.

## Possible fixes

There are workarounds, i.e. to always give parameters in a dict with their name, the names triggering a retrace.
```

### Standalone code to reproduce the issue

```shell
Link to colab with the code below: https://colab.research.google.com/drive/1z26ZIoiFEBO9icGJ5bYusu4XY5u-vpkG?usp=sharing


import tensorflow as tf

x1 = tf.Variable(2.0)
x2 = tf.Variable(4.0)


def f():
    res = x1 + x2 ** 2 / 2
    return res


def grad(param):
    with tf.GradientTape(watch_accessed_variables=False) as tape:
        tape.watch(param)
        value = f()
    return tape.gradient(value, param)

jitted_grad = tf.function(grad)

y1 = grad(x1)
y1_jit = jitted_grad(x1)
assert abs(y1 - 1.0) < 1e-5  # because d x1 / dx1 = 1
assert abs(y1_jit - 1.0) < 1e-5
y2 = grad(x2)
y2_jit = jitted_grad(x2)
print(f"y2: {y2}, should be 4")  # but is 1 because it uses the derivative of x1
print(f"y2_jit: {y2_jit}, should also be 4")  # but is 1 because it uses the derivative of x1
assert abs(y2 - 4.0) < 1e-5  # because d / dx x**2/2 = x -> 4
assert abs(y2_jit - 4.0) < 1e-5  # fails!
```
```


### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow [v2.8](https://colab.research.google.com/gist/tilakrayal/8935c59e2233d0896d153a2071d36fe9/2-8gradient_tape_with_function_and_variable_broken.ipynb), [v2.9](https://colab.research.google.com/gist/tilakrayal/edb62f126ef28301034693a1ea2be901/2-9gradient_tape_with_function_and_variable_broken.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/d72b807fe4163d15e5e6f88dd508fe28/2-11gradient_tape_with_function_and_variable_broken.ipynb). Kindly find the gist of it here.

**Comment by jonas-eschle** [comment]
@sachinprasadhs I do not want to sound alarmist, and I am aware that there are many things going on in TF, but this is a quite serious bug: gradients of jitted GradientTape will be completetly wrong! This will severely impact neural network training for anything that is outside of the standard trainings. It still exists even in the nightlies (which are 2.14.x).

Is there any way of putting some more importance to this issue? (again, not as in "it's my issue, please resolve", but I do honestly believe that it is a deeper issue with the design that probably now causes thousands of hard-to-recover bugs in peoples code)

**Comment by jonas-eschle** [comment]
@sachinprasadhs just a slight ping, as mentioned, this is actually a HUGE issue that people surely already ran into but maybe haven't been able to notice. Don't you think so?
